File: src/modifiers/digitalCrownAccessory.ts
import type { ReactNode } from 'react';

import { createModifier } from './createModifier';

export interface DigitalCrownAccessoryParams {
  /// View placed alongside the Digital Crown while the crown is in use
  /// (e.g. a value readout). Hoisted into a hidden content slot.
  content: ReactNode;
}

/// SwiftUI `.digitalCrownAccessory { … }` — shows an accessory view next to
/// the Digital Crown during interaction. Available on watchOS 9 (the
/// deployment floor).
export function digitalCrownAccessory(
  content: ReactNode
): ReturnType<typeof createModifier>;
export function digitalCrownAccessory(
  params: DigitalCrownAccessoryParams
): ReturnType<typeof createModifier>;
export function digitalCrownAccessory(
  arg: ReactNode | DigitalCrownAccessoryParams
) {
  if (arg != null && typeof arg === 'object' && 'content' in arg) {
    return createModifier('digitalCrownAccessory', arg);
  }
  return createModifier('digitalCrownAccessory', { content: arg as ReactNode });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/DigitalCrownAccessory.swift
import SwiftUI

/// SwiftUI `.digitalCrownAccessory { … }`. Shows an accessory view next to the
/// Digital Crown while it's in use (e.g. a value readout). The accessory body
/// is passed as modifier content from JS and resolved through `ctx.content`;
/// when no content is provided the view is returned unchanged.
///
/// Available on watchOS 9, the deployment floor, so no availability gating is
/// required.
enum RNWDigitalCrownAccessoryModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("digitalCrownAccessory") { view, params, ctx in
            guard let body = ctx.content(params.string("content")) else {
                return view
            }
            return AnyView(view.digitalCrownAccessory { body })
        }
    }
}
